scripts/json2synapse.py

```diff
@@ -0,0 +1,217 @@
+ngs = 'https://raw.githubusercontent.com/Sage-Bionetworks/synapseAnnotations/master/synapseAnnotations/data/ngs.json'
+tool = 'https://raw.githubusercontent.com/Sage-Bionetworks/synapseAnnotations/master/synapseAnnotations/data/tool.json'
+toolExtended = 'https://raw.githubusercontent.com/Sage-Bionetworks/synapseAnnotations/master/synapseAnnotations/data/toolExtended.json'
+
+paths = [analysis, sageCommunity, array, cancer, dhart, experimentalData, genie, neuro, neurofibromatosis, ngs, tool, toolExtended]
+names = ['analysis', 'sageCommunity', 'array', 'cancer', 'dhart', 'experimentalData', 'genie', 'neuro', 'neurofibromatosis', 'ngs', 'tool', 'toolExtended']
+
+
+all_modules = []
+
+tableSynId = "syn10242922"
+key = ["key", "value"]
+
+
+def json2flatten(path, module):
+    # fetch and read raw json objects from its' github url and decode the json object to its raw format
+    json_record = pandas.read_json(path)
+
+    # grab annotations with empty enumValue lists i.e. don't require normalization and structure their schema
+    empty_vals = json_record.loc[json_record.enumValues.str.len() == 0]
+    empty_vals = empty_vals.drop('enumValues', axis=1)
+
+    empty_vals['enumValues_description'] = ""
+    empty_vals['enumValues_source'] = ""
+    empty_vals['enumValues_value'] = ""
+    empty_vals['module'] = module
+
+    empty_vals.set_index(empty_vals['name'], inplace=True)
+
+    # for each value list object
+    flatten_vals = []
+    json_record = json_record.loc[json_record.enumValues.str.len() > 0]
+    json_record.reset_index(inplace=True)
+
+    for i, jsn in enumerate(json_record['enumValues']):
+        df = pandas.io.json.json_normalize(json_record['enumValues'][i])
+        # re-name columns to match table on synapse schema
+        df = df.rename(columns={'value': 'enumValues_value', 'description': 'enumValues_description',
+                                'source': 'enumValues_source'})
+
+        # grab key information in its row, expand it by values dimention and append its key-columns to flattened values
+        rows = json_record.loc[i:i, json_record.columns != 'enumValues']
+        repeats = pandas.concat([rows] * len(df.index))
+        repeats.set_index(df.index, inplace=True)
+        flatten_df = pandas.concat([repeats, df], axis=1)
+        # append module category / annotating the annotations for module filtering
+        flatten_df['module'] = module
+        flatten_df.set_index(flatten_df['name'], inplace=True)
+
+        flatten_vals.append(flatten_df)
+
+    flatten_vals.append(empty_vals)
+
+    return flatten_vals
+
+
+def rightJoin(df1, df2, key, suffixes):
+    # rows in df2 that are not in df1 based on unique keys (can be a list) 
+    df3 = pandas.merge(df1, df2, on=key, how='right', indicator='Exist', suffixes=suffixes)
+    df3['Exist'] = numpy.where(df3.Exist == 'both', True, False)
+    return df3
+
+
+def outerJoin(df1, df2):
```

Comment:
Outer join will essentially identify rows in df1 that are not in df2. 
https://pandas.pydata.org/pandas-docs/stable/merging.html
```python 
In [179]: pandas.merge(df1, df2, how='outer', indicator=True).query('_merge == "left_only"').drop(['_merge'], axis=1)
Out[]:
   col1  col2 key1 key2 key3
0     1     2    a    x    (
2     1     2    c    z    )
```
I use it a couple times to identify the rows changed in current table on synapse (given both case 1 and 2), I remove those rows. Then I add the overall changed and new rows and update the table.